### DISCUSSION TITLE
Do vendor files need to be committed ?

### DISCUSSION THREAD

**Discussion by Ig0r-M-magic42** [rectorphp/rector#7339] [Q&A] [answered]
## Question
- Do vendor files need to be committed ?
## Cons seen from this
- Required packages by rector which are installed in `vendor/rector/rector/vendor` might collide in versions with packages installed in `vendor`, although composer will not complain but autoload might be messed up.

## Example
Using version 0.12.19 there was an issue with phpdoc due to this requirement `"phpstan/phpdoc-parser": "<1.2"` and [PHPStan\PhpDocParser\Parser\TokenIterator::consumeTokenValue()](https://github.com/phpstan/phpdoc-parser/blob/aa111aa1fecbc436f01e5439d88906bd3b612027/src/Parser/TokenIterator.php#L100) was implemented only in 1.4.0
![image](https://user-images.githubusercontent.com/22006850/182700222-b3887c36-f259-47f7-8912-0559bbde2347.png)

<details>
<summary>Stack Trace</summary>

```
Fatal error: Uncaught Error: Call to undefined method PHPStan\PhpDocParser\Parser\TokenIterator::consumeTokenValue() in /vagrant/vendor/phpstan/phpdoc-parser/src/Parser/PhpDocParser.php:435
Stack trace:
#0 /vagrant/vendor/phpstan/phpdoc-parser/src/Parser/PhpDocParser.php(212): PHPStan\PhpDocParser\Parser\PhpDocParser->parseTypeAliasImportTagValue()
#1 /vagrant/vendor/phpstan/phpdoc-parser/src/Parser/PhpDocParser.php(114): PHPStan\PhpDocParser\Parser\PhpDocParser->parseTagValue()
#2 /vagrant/vendor/phpstan/phpdoc-parser/src/Parser/PhpDocParser.php(75): PHPStan\PhpDocParser\Parser\PhpDocParser->parseTag()
#3 /vagrant/vendor/phpstan/phpdoc-parser/src/Parser/PhpDocParser.php(48): PHPStan\PhpDocParser\Parser\PhpDocParser->parseChild()
#4 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/PhpDoc/PhpDocStringResolver.php(24): PHPStan\PhpDocParser\Parser\PhpDocParser->parse()
#5 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/FileTypeMapper.php(438): PHPStan\PhpDoc\PhpDocStringResolver->resolve()
#6 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/FileTypeMapper.php(242): PHPStan\Type\FileTypeMapper->getTypeAliasesMap()
#7 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/FileTypeMapper.php(457): PHPStan\Type\FileTypeMapper->PHPStan\Type\{closure}()
#8 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/FileTypeMapper.php(468): PHPStan\Type\FileTypeMapper->processNodes()
#9 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/FileTypeMapper.php(463): PHPStan\Type\FileTypeMapper->processNodes()
#10 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/FileTypeMapper.php(468): PHPStan\Type\FileTypeMapper->processNodes()
#11 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/FileTypeMapper.php(423): PHPStan\Type\FileTypeMapper->processNodes()
#12 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/FileTypeMapper.php(181): PHPStan\Type\FileTypeMapper->createFilePhpDocMap()
#13 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/FileTypeMapper.php(168): PHPStan\Type\FileTypeMapper->createResolvedPhpDocMap()
#14 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Type/FileTypeMapper.php(75): PHPStan\Type\FileTypeMapper->getResolvedPhpDocMap()
#15 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ClassReflection.php(849): PHPStan\Type\FileTypeMapper->getResolvedPhpDoc()
#16 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ClassReflection.php(861): PHPStan\Reflection\ClassReflection->getResolvedPhpDoc()
#17 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ClassReflection.php(853): PHPStan\Reflection\ClassReflection->getExtendsTags()
#18 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ClassReflection.php(164): PHPStan\Reflection\ClassReflection->getFirstExtendsTag()
#19 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ClassReflection.php(577): PHPStan\Reflection\ClassReflection->getParentClass()
#20 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ReflectionProvider/ClassBlacklistReflectionProvider.php(49): PHPStan\Reflection\ClassReflection->getParentClassesNames()
#21 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ReflectionProvider/ChainReflectionProvider.php(35): PHPStan\Reflection\ReflectionProvider\ClassBlacklistReflectionProvider->hasClass()
#22 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ReflectionProvider/MemoizingReflectionProvider.php(39): PHPStan\Reflection\ReflectionProvider\ChainReflectionProvider->getClass()
#23 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ClassReflection.php(175): PHPStan\Reflection\ReflectionProvider\MemoizingReflectionProvider->getClass()
#24 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ClassReflection.php(577): PHPStan\Reflection\ClassReflection->getParentClass()
#25 phar:///vagrant/vendor/rector/rector/vendor/phpstan/phpstan/phpstan.phar/src/Reflection/ReflectionProvider/ClassBlacklistReflectionProvider.php(49): PHPStan\Reflection\ClassReflection->getParentClassesNames()
```
## Note
- This happened when updating to magento 2.4.4

**Comment by TomasVotruba** [comment, marked as answer]
The vendor is needed, so Rector can be installed with own dependency tree on old projects. E.g. with Symfony 2.8.
It is often scoped to avoid conflicts you describe: https://getrector.org/blog/prefixed-rector-by-default

The package you've came across is the one that is missing `composer.json`. It's part of PHPStan, so an implicit package. 
It's now handled: https://github.com/rectorphp/rector-src/blob/61b6e86d17923a29a7dad1530a9dd82bc3bf5ee2/build/target-repository/composer.json#L23

The 0.12 version is light-years old. 

Does the probleam remain with 0.13.10?

**Reply by Ig0r-M-magic42** [reply]
Thanks @TomasVotruba for sharing with the https://getrector.org/blog/prefixed-rector-by-default which explains a lot of things to me.

The reason why i have version 0.12.19 is due to it being a dependency of `magento/magento-coding-standard`, which i will try to update.
![image](https://user-images.githubusercontent.com/22006850/182796526-bd2d8e84-f5d1-4af0-a47e-ad083d8e5f72.png)

Questions in regards of the decision behind committing vendor files, have you considering compiling a .phar that can be installed globally which will have all it's vendor composer dependencies thus having 2 approaches to install it:
- Either global install rector or download the .phar which will cover the need to support older version of different frameworks e.g. Symfony 2.8
- Installed as a composer dev dependency which will install it's dependencies in the `vendor` folder whilst not having `rector/rector/vendor` files???

**Reply by TomasVotruba** [reply]
The phar does not solve the dependency issue. Also phar blocks you from modifying the Rector code for testing, debugging and writing your own rules.

The magento package should never lock the top version of the dependency. Unless temporary hotfix.

---

Basically phar is the same as the code, just in a zip file :) You still run it like `vendor/bin/rector`, but the same classes are autoloaded. Phar makes sense in an Enterprise version with extra features, where you need to hide sensitive code used e.g. for authentication. That should not be modified... working on that :)

**Reply by Ig0r-M-magic42** [reply]
@TomasVotruba and i do agree that .phar comes with it's limitations as you've mentioned those above, though i do not envision everyone which installs rector with the purpose of debugging the rector code itself or writing own rules, by providing multiple options it allows developer to pick the needs that suits best for their project.

Regarding the vendor files being commited, is it possible to downgrade the priority of the loader to look in `rector/rector/vendor` after `vendor` ?

**Reply by TomasVotruba** [reply]
We use static reflection approach from PHPStan. See https://phpstan.org/user-guide/autoloading

**Reply by Ig0r-M-magic42** [reply]
Thanks for the answer, no further questions at this point.
